### DISCUSSION TITLE
X features in ".predict"

### DISCUSSION THREAD

**Discussion by Ehsan86** [Nixtla/mlforecast#349] [Q&A] [answered]
In the code below, the pipeline transforms the X-featurs and then fit the model. My questions is, does it also automatically trasnform the input features in ".predict"? or i have to transform the features first using the preprocessor and then pass it to ".predict"?
When I print the input features in prediction using callback module, it shows the untransformed version of features. 
* note: here when i say transform i refer to preprocessor transformation part to the lag transformation.

```
preprocessor = ColumnTransformer(
    transformers=[
        ('passthrough', 'passthrough', features_remaining),
        ('robust_scaler', RobustScaler(), feature_robust_scaler),
        ('one_hot_encoder', OneHotEncoder(), features_hot_encoder),   
    ]
)
```

```
pipeline = make_pipeline(preprocessor, 
                         RandomForestRegressor(n_estimators=100, min_samples_split=10, min_samples_leaf=4))

forecast_model = MLForecast(
    models={"model": pipeline},
    freq="30T",
    lags=[1, 35, 47, 48, 49, 335, 336, 337, 338, 672],
     lag_transforms={  
        1: [RollingMean(window_size=12), 
            RollingMean(window_size=48), 
            RollingMean(window_size=336), 
            RollingMean(window_size=672)],
    }
)
```

```
def inspect_input(new_x):
    """Displays the model inputs to inspect them"""
    display(new_x)
    return new_x
```

```
forecast_model.fit(df_train, static_features=["unique_id"], fitted=True, dropna=False)
forecast_model.predict(h=2, before_predict_callback=inspect_input, `X_df=df_test.drop(columns=["y"]))
```

**Comment by jmoralez** [comment]
Hey. The callback returns what gets passed to what mlforecast interprets as the model (in this case the pipeline), so the features will be transformed during predict, you just can't see them with the callback. You could add a step to the pipeline that prints them before passing them to the real model if you want to see for yourself.

**Reply by Ehsan86** [reply]
Thank you @jmoralez for your explanation. May I know in this case, where and how I should add the print statement in the pipeline stage. I appreciate it if you would assist me in this.

**Reply by jmoralez** [reply, marked as answer]
You can add a simple estimator that just prints, e.g.
```python
from sklearn.base import BaseEstimator, TransformerMixin

class PrintEstimator(BaseEstimator, TransformerMixin):
    def fit(self, X, y):
        return self

    def transform(self, X, y=None):
        print('Features going into the model: ', X)
        return X

pipeline = make_pipeline(
    preprocessor,
    PrintEstimator(),
    RandomForestRegressor(n_estimators=100, min_samples_split=10, min_samples_leaf=4)
)
```
